### PR TITLE
Lint and format with eslint and prettier

### PR DESCRIPTION
**Pull request by boxfoot**
Configures eslint and prettier to run on staged code as a pre-commit hook, and also applies current rules to all existing code.  Addresses #218 & #455.

Note @wcjordan that I started with the default configuration for just about everything (except swapping from double quotes to single quotes, since that's the previous practice).   IME it takes some time to get used to prettier's formatting, especially for some long lines.  Please look through the PR and flag if any of the styling choices are particularly grating for you and I can check if there's an option to control that.

Also note that I ran this on top of PR #519 , assuming that would be merged first and to avoid conflicts.  I am happy to rebase and rerun the code formatting if you prefer.

## How Has This Been Tested?
All code changes automated through eslint and prettier, so there shouldn't be any functionality changes.  In addition, ran all unit tests.

## Types of changes
- [x] Code formatting/clean-up

## Checklist:
- [x] My code follows the code style of this project.
- [x] ~My change requires a change to the documentation.~
- [x] ~I have updated the documentation accordingly.~
- [x] I have read the **CONTRIBUTING** document.

### CODE REVIEW

**Comment by boxfoot** [comment]
This PR fails the Travis build on node@8, since husky requires node>=10.   Given that [node 8 is now EOL](https://github.com/nodejs/Release) would it be OK to drop that Travis build target?

**Comment by wcjordan** [comment]
Dropping node 8 sounds good.  Let's do 10, 12, & 13 for now

File: package.json
Comment (wcjordan) [inline package.json]: Are these used?

File: package.json
Comment (wcjordan) [inline package.json]: Let's pin exact for the dev dependencies so there's no risk they break

File: package.json
Comment (boxfoot) [reply package.json]: These are used -- although not actively imported.   Since this repo uses type annotations in most JS files, my IDE (vscode) (and I expect others as well) can check types.   The repo is actually in very good shape type-wise, but there are a ton of errors in the test files b/c there are no type definitions for `mocha` and `chai`, which is why I added them to the repo.  Adding these dependencies silences very many false errors, and lets the type checking actually zoom in on real code problems to fix.
(These dependencies are just the type definitions from definitely-typed, no actual functional code).

Would you prefer that I remove them?

**Comment by boxfoot** [comment]
> Dropping node 8 sounds good. Let's do 10, 12, & 13 for now

OK.   What do you think of pegging to latest stable (which is currently 13) and latest LTS (currently 12) plus 10 for good measure?  I did that in my last commit, and all pass.  Or if you prefer that I hard code to 10, 12, and 13, LMK and I'll update that.

File: package.json
Comment (wcjordan) [reply package.json]: Ah makes sense.  If it's helpful to you feel free to leave them.  They're not harming anything.

**Comment by wcjordan** [comment]
That approach to travisci seems fine to me.  Thanks!

**Comment by wcjordan** [comment]
Holding off merging this while we work out some issues w/ how removing hasteResolvers affects our NPM deployment.

**Comment by wcjordan** [comment]
@pradeepnschrodinger is this something we could see about merging / re-applying now that we've completed the imports work?

**Comment by pradeepnschrodinger** [comment]
> @pradeepnschrodinger is this something we could see about merging / re-applying now that we've completed the imports work?

Yea. We can get this merged to master.

@boxfoot , could you reapply the rules and fix the merge conflicts?

**Comment by boxfoot** [comment]
Sure thing, will do.

**Comment by boxfoot** [comment]
@pradeepnschrodinger @wcjordan rebased onto master and formatting reapplied.  LMK how this looks.

**Review by wcjordan — APPROVED** [review]
Thanks, this looks good to me.  @pradeepnschrodinger do you have any concerns?

Since we have some large refactors currently in flight, I think the best way to merge this may be to create a new PR with just the non-automated changes (e.g. package.json, .travis.yml, *rc.yaml) and merge that.  Then we can apply the automated formatting after the refactors are merged.  Thoughts?

**Comment by pradeepnschrodinger** [comment]
> Since we have some large refactors currently in flight, I think the best way to merge this may be to create a new PR with just the non-automated changes (e.g. package.json, .travis.yml, *rc.yaml) and merge that. Then we can apply the automated formatting after the refactors are merged. Thoughts?

Hmm, yea having another PR to review and apply the formatted changes sounds good. 

Sorry for requesting again, @boxfoot, could you open up another PR with just the config changes?

**Comment by pradeepnschrodinger** [comment]
> > Since we have some large refactors currently in flight, I think the best way to merge this may be to create a new PR with just the non-automated changes (e.g. package.json, .travis.yml, *rc.yaml) and merge that. Then we can apply the automated formatting after the refactors are merged. Thoughts?
> 
> Hmm, yea having another PR to review and apply the formatted changes sounds good.
> 
> Sorry for requesting again, @boxfoot, could you open up another PR with just the config changes?

ping: @boxfoot

**Comment by boxfoot** [comment]
@pradeepnschrodinger sorry for the delay - lots going on in life and work that were pulling me away.   Updated to just have the config.

When you're ready to do an autoformat commit, you can run `npx prettier --write '**/*{js,html,md}'` until then, you should expect your PRs to be a bit messy because they'll mix autoformatting with meaningful changes, making the meaningful changes harder to see.

One strategy you can use to mitigate that it asking contributors to have their first commit be an autoformat of files touched in their PR, so that you can use the github review feature of looking at meaningful changes separately during review.

**Comment by wcjordan** [comment]
Great, thank you @boxfoot !

@pradeepnschrodinger I recommend we merge this.  Then immediately run the formatter with
npx prettier --write '**/*{js,html,md}'
and merge that as a 2nd commit.  Does that sound good to you?

**Comment by pradeepnschrodinger** [comment]
@boxfoot , thanks!
> One strategy you can use to mitigate that it asking contributors to have their first commit be an autoformat of files touched in their PR, so that you can use the github review feature of looking at meaningful changes separately during review.

That sounds good.

> @pradeepnschrodinger I recommend we merge this. Then immediately run the formatter with
> npx prettier --write '**/*{js,html,md}'
> and merge that as a 2nd commit. Does that sound good to you?

@wcjordan , I think we should hold off on committing the autoformatted changes until the other refactors take place.
We can merge this PR meanwhile. Does that sound okay?

**Comment by wcjordan** [comment]
Sure, that works for me.  Just need to be a bit more diligent to ensure we don't get a lot of noise on commits from the autoformatter running on edited files.

**Pull request merged**
